src/vmm/src/vmm_config/drive.rs

```diff
@@ -70,109 +81,114 @@ pub struct BlockDeviceConfig {
+                    if self.has_root_block {
+                        return Err(DriveError::RootBlockDeviceAlreadyAdded);
+                    } else {
+                        // One of the non-root blocks is becoming root.
+                        self.has_root_block = true;
+
+                        // Make sure the root device is on the first position.
+                        self.list.swap(0, index);
+                        // Block config to be updated has moved to first position.
+                        index = 0;
+                    }
+                }
+                // Update the slot with the new block.
+                self.list[index] = block_dev;
             }
-        } else {
-            self.config_list.push_back(block_device_config);
         }
-
         Ok(())
     }
 
-    /// Updates a Block Device Config. The update fails if it would result in two
-    /// root block devices.
-    fn update(&mut self, mut index: usize, new_config: BlockDeviceConfig) -> Result<()> {
-        // Check if the path exists
-        if !new_config.path_on_host.exists() {
+    /// Creates a Block device from a BlockDeviceConfig.
+    pub fn create_block(block_device_config: BlockDeviceConfig) -> Result<Block> {
+        // check if the path exists
+        if !block_device_config.path_on_host.exists() {
             return Err(DriveError::InvalidBlockDevicePath);
         }
 
-        // Check if the root block device is being updated.
-        if self.config_list[index].is_root_device {
-            self.has_root_block = new_config.is_root_device;
-        } else if new_config.is_root_device {
-            // Check if a second root block device is being added.
-            if self.has_root_block {
-                return Err(DriveError::RootBlockDeviceAlreadyAdded);
-            } else {
-                // One of the non-root blocks is becoming root.
-                self.has_root_block = true;
-
-                // Make sure the root device is on the first position.
-                self.config_list.swap(0, index);
-                // Block config to be updated has moved to first position.
-                index = 0;
-            }
-        }
-        // Update the config.
-        self.config_list[index] = new_config;
-
-        Ok(())
+        // Add the block device from file.
+        let block_file = OpenOptions::new()
+            .read(true)
+            .write(!block_device_config.is_read_only)
+            .open(&block_device_config.path_on_host)
+            .map_err(DriveError::OpenBlockDevice)?;
+
+        let rate_limiter = block_device_config
+            .rate_limiter
+            .map(super::RateLimiterConfig::try_into)
+            .transpose()
+            .map_err(DriveError::CreateRateLimiter)?;
+
+        // Create and return the Block device
+        devices::virtio::Block::new(
+            block_device_config.drive_id,
+            block_file,
```

Comment:
WDYT about passing the block file path to the device instead of the file? This way we wouldn't need an outside config with the path when saving/restoring the device.
This can be done in the block serialization PR.

Reply:
Sure, I don't have any strong reason against it.